refactor: move `DataSource` to `datafusion-datasource`

logan-keede opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Part of #14444.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?
refactor of `DataSource`
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?
not completely, only `cargo check` work for now

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?
yes, but only if they are building from source.
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

alamb commented:
So exciting...

logan-keede commented:
@alamb any Idea on what to do with unit tests in `physical-plan`? I cant think of anything except pulling functions that depend on the component that I moved to `core_integration` tests or maybe to `datasource` crate. is there a better method?

alamb commented:
> @alamb any Idea on what to do with unit tests in `physical-plan`? I cant think of anything except pulling functions that depend on the component that I moved, to `core_integration` tests or maybe to `datasource` crate. is there a better method?

I looked at the errors, and most of them appear to be related to the tests in physical_plan using `MemoryExec` (well now DataSource with a MemoryExec).

Some other ideas:
1. Add a `dev-dependency` between datafusion-physical-plan` and `datafusion-datasource`
2.  (better) Add a mock `ExecutionPlan` only for testing in `physical_plan` -- this would end up with some amount of duplicate code with DataSourceExec but I think that would be ok for testing. 
3. Move the tests, as you say, into tests in `core_integration` or something

alamb commented:
BTW thank you for working on this -- I suggest we try the "make a mock ExecutionPlan" approach -- that would be the cleanest I think and would be a good way for you to learn more about how ExecutionPlans worked

I can try and find time to sketch it out if needed

logan-keede commented:
> BTW thank you for working on this -- I suggest we try the "make a mock ExecutionPlan" approach -- that would be the cleanest I think and would be a good way for you to learn more about how ExecutionPlans worked

Thanks for the suggestions. I will try to make `MockExecutionPlan`.

logan-keede commented:
@alamb, I have implemented a MockMemorySourceConfig and MockDataSource which is essentially same as what was previously in physical_plan but now for tests only. You can find the implementation [here](https://github.com/logan-keede/datafusion/tree/explore_Mock). The tests are passing on my PC, although there are some warnings, so cleaning up that part is still pending.

I think this is different from what you were suggesting. It would be pretty helpful if you can point me in the right direction.

Thanks, 
Logan

alamb commented:
> @alamb, I have implemented a MockMemorySourceConfig and MockDataSource which is essentially same as what was previously in physical_plan but now for tests only. You can find the implementation [here](https://github.com/logan-keede/datafusion/tree/explore_Mock). The tests are passing on my PC, although there are some warnings, so cleaning up that part is still pending.
> 
> I think this is different from what you were suggesting. It would be pretty helpful if you can point me in the right direction.
> 
> Thanks, Logan

I think this sounds reasonable -- I was just saying that the split between `MockMemorySourceConfig` and `MockDataSource` is likely not necessary.  There is a split in the current `DataSourceConfig` and `DataSourceExec` to permit sharing between difefrent data sources

logan-keede commented:
> I think this sounds reasonable -- I was just saying that the split between `MockMemorySourceConfig` and `MockDataSource` is likely not necessary.

It does look that way, I will try to make it one Mock struct.

logan-keede commented:
@alamb What do we want to do with benches? (they are causing circular dependency)
except that I think this is ready for review.

alamb commented:
> @alamb What do we want to do with benches? (they are causing circular dependency) except that I think this is ready for review.

I think we should move the benches into `datafusion/core/benches`

alamb commented:
Thank you very much for this PR @logan-keede  -- I will try and review it later today (I am out of the office this week so I have only limited time to review PRs)

alamb commented on datafusion-cli/src/functions.rs:
```diff
@@ -28,10 +28,10 @@ use arrow::record_batch::RecordBatch;
 use arrow::util::pretty::pretty_format_batches;
 use datafusion::catalog::{Session, TableFunctionImpl};
 use datafusion::common::{plan_err, Column};
+use datafusion::datasource::memory::MemorySourceConfig;
```
![giphy](https://github.com/user-attachments/assets/b8ab0fec-8c3e-46c2-8524-c30d46e5fa4d)

alamb commented on datafusion/core/Cargo.toml:
```diff
@@ -222,3 +222,7 @@ required-features = ["nested_expressions"]
 name = "dataframe"
+
+[[bench]]
+harness = false
```
👍

alamb commented on datafusion/datasource/src/memory.rs:
```diff
@@ -0,0 +1,926 @@
+
+/// Data source configuration for reading in-memory batches of data
+#[derive(Clone)]
+pub struct MemorySourceConfig {
```
I am just confirming that the plan is that datasource will have MemorySourceConfig / MemorySource and that we will move the other format specific things (like parquet) to their own crates like `datafusion-datasource-parquet` ?

logan-keede replied:
Yes, That's what I had in my mind so far. However, now that I think about it, I don't think format-specific crate(except parquet) will have more than 2(fileformat, Source+mod) files, If we are going to keep a separate crate for them, maybe we can keep a separate crate for `MemorySourceConfig`(1 file, Source + mod.rs ) too. What do you think?

alamb commented on datafusion/physical-plan/src/joins/nested_loop_join.rs:
```diff
@@ -1030,8 +1030,8 @@ impl EmbeddedProjection for NestedLoopJoinExec {
-    use crate::memory::MemorySourceConfig;
-    use crate::source::DataSourceExec;
+    use crate::test::MockMemorySourceConfig;
+    // use crate::test::MockMemorySourceConfig;
```
left over?

logan-keede replied:
yes, at a few more places too. I think I have cleaned this out with latest commit.

alamb commented on datafusion/physical-plan/src/joins/test_utils.rs:
```diff
@@ -23,9 +23,9 @@ use crate::joins::utils::{JoinFilter, JoinOn};
 use crate::repartition::RepartitionExec;
-use crate::source::DataSourceExec;
+use crate::test::MockMemorySourceConfig;
+// use crate::test::MockMemorySourceConfig;
```
```suggestion
```

alamb commented on datafusion/physical-plan/src/test.rs:
```diff
@@ -17,27 +17,347 @@
 pub mod exec;
 
+#[derive(Clone, Debug)]
+pub struct MockMemorySourceConfig {
```
Some suggestions:
1. Document that this implements an in memory `DataSource` for testing purposes (and maybe we can add a link to the real memory source) and that it duplicates much of the `MemorySourceConfig` to avoid a dependency between physical-plan and datasource
2. Rename it to something like `TestMemoryExec`  as this is not a config but rather something that `impl ExecutionPlan` and is used as a source of data in the tests

mertak-synnada commented on datafusion/datasource/src/memory.rs:
```diff
@@ -0,0 +1,926 @@
+}
+
+#[cfg(test)]
+mod memory_exec_tests {
```
Nit: This looks forgotten. While moving the tests, I think we can rename the module as `memory_source_tests` since there's no `MemoryExec`

mertak-synnada reviewed (approved):
LGTM! Thank you for all the hard work. I’ve added a few optional minor comments

logan-keede commented:
@alamb @mertak-synnada TYSM for the review!! I think I have addressed your suggestions in the latest commit. Let me know if they look good to you (especially the documentation part) or if you have anything else.

alamb commented on datafusion/physical-plan/src/test.rs:
```diff
@@ -49,8 +49,15 @@ use futures::{Future, FutureExt};
 pub mod exec;
 
+/// `TestMemoryExec` is a mock equivalent to [`MemorySourceConfig`] with [`ExecutionPlan`] implemented for testing.
+/// i.e. It has some but not all the functionality of [`MemorySourceConfig`].
```
👍

alamb commented:
Thanks @logan-keede and @mertak-synnada  -- this looks great and thanks (again) for driving the plan forward
